api/src/main/java/org/apache/iceberg/transforms/ProjectionUtil.java

```diff
@@ -231,7 +231,9 @@ static <S, T> UnboundPredicate<T> truncateArrayStrict(
   static <T> UnboundPredicate<T> projectTransformPredicate(
       Transform<?, T> transform, String partitionName, BoundPredicate<?> pred) {
     if (pred.term() instanceof BoundTransform
-        && transform.equals(((BoundTransform<?, ?>) pred.term()).transform())) {
+        && transform
```

Comment:
I had to change this to make `testProjectionNames` pass. In the case of `hour` transform, we'll be comparing `transforms` which is an instance of `Timestamps` with `Hours`. The `toString()` both produce `hour`.

Reply:
We should be able to roll this back once we deprecate the old `apply` method in favor of the `bind` approach. Then we don't need to keep two classes around for those.